### PR TITLE
Add CCT Classifier Training Support

### PR DESCRIPTION
**Pull request by runwangdl**
## Description  
This PR focuses on fine-tuning the final GEMM operator in CCT and preparing for future training support. It includes new test cases, refinements in multiple operator parsers and `computeShape` logic, and improved compatibility for the `layernorm` operator to support `layernormgrad` in training in the future.

## Added  
- New test cases:  
  - `testTrainCCT/CCT_GEMM_Weight_Bias_1_16_16_8`  
  - `testFloatReduceSum`  
  - `testFloatSoftmaxGrad`  
  - `testFloatSoftmaxCrossEntropy`  
  - `testFloatSoftmaxCrossEntropyGrad`  
- New kernels
  - `SoftmaxCrossEntropy`
  - `SoftmaxCrossEntropyGrad`
  - `SoftmaxGrad`
  - `ReduceSum`
- Refinements in operator parsers and `computeShape` logic for:  
  - `Softmax`  
  - `Mul`  
  - `Reducesum`  

## Changed  
- Enhanced `layernorm` operator to support three outputs (`layernormout`, `mean`, `std`) for compatibility with training-related `layernormgrad` in the future.  

## PR Merge Checklist

1. [x] The PR is rebased on the latest `devel` commit and pointing to `devel`.
2. [x] Your PR reviewed and approved.
3. [x] All checks are passing.
4. [x] The `CHANGELOG.md` file has been updated.
5. [x] If the docker was modified, change back its link after review.

### CODE REVIEW

File: Deeploy/Targets/Generic/Parsers.py [hunk 1]
```diff
@@ -2322,3 +2381,64 @@ def parseNodeCtxt(self,
...
+
+        logits = ctxt.lookup(node.inputs[0].name)
+        labels = ctxt.lookup(node.inputs[1].name)
+        loss = ctxt.lookup(node.outputs[0].name)
+        log_prob = ctxt.lookup(node.outputs[1].name)
+        self.operatorRepresentation['logits'] = logits.name
+        self.operatorRepresentation['labels'] = labels.name
+        self.operatorRepresentation['loss'] = loss.name
+        self.operatorRepresentation['log_prob'] = log_prob.name
+        self.operatorRepresentation['batch'] = np.prod(logits.shape[0])
```
Comment (Victor-Jung) [inline Deeploy/Targets/Generic/Parsers.py:2410]: Why are you doing the product of a shape at index 0? Shouldn't this be a scalar?

File: Deeploy/Targets/Generic/Parsers.py [hunk 2]
```diff
@@ -2322,3 +2381,64 @@ def parseNodeCtxt(self,
...
+        self.operatorRepresentation['num_classes'] = np.prod(logits.shape[1])
```
Comment (Victor-Jung) [inline Deeploy/Targets/Generic/Parsers.py:2411]: Same question here.

File: Deeploy/Targets/Generic/Parsers.py [hunk 3]
```diff
@@ -2322,3 +2381,64 @@ def parseNodeCtxt(self,
...
+        log_prob = ctxt.lookup(node.inputs[1].name)
+        labels = ctxt.lookup(node.inputs[2].name)
+        grad = ctxt.lookup(node.outputs[0].name)
+
+        self.operatorRepresentation['loss_grad'] = loss_grad.name
+        self.operatorRepresentation['log_prob'] = log_prob.name
+        self.operatorRepresentation['labels'] = labels.name
+        self.operatorRepresentation['grad'] = grad.name
+        self.operatorRepresentation['batch'] = np.prod(log_prob.shape[0])
+        self.operatorRepresentation['num_classes'] = np.prod(log_prob.shape[1])
```
Comment (Victor-Jung) [inline Deeploy/Targets/Generic/Parsers.py:2442]: Same question as before here.

File: Deeploy/Targets/PULPOpen/Templates/SoftmaxCrossEntropyLossTemplate.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+# ----------------------------------------------------------------------
+#
+# File: PadTemplate.py
```
Comment (Victor-Jung) [inline Deeploy/Targets/PULPOpen/Templates/SoftmaxCrossEntropyLossTemplate.py:3]: NIT: Wrong name here

File: .github/workflows/CI.yml [hunk 5]
```diff
@@ -270,6 +270,11 @@ jobs:
         testFloatMul
         Quant
         Dequant
+        testTrainCCT/CCT_GEMM_Weight_Bias_1_16_16_8
```
Comment (Victor-Jung) [inline .github/workflows/CI.yml:273]: This is a model, not a kernel. It goes into the `siracusa-models` job.

**Review by Victor-Jung — COMMENTED** [review]
LGTM overall, a few minor comments.

File: Deeploy/Targets/Generic/Parsers.py (hunk 1 above)
Comment (runwangdl) [reply Deeploy/Targets/Generic/Parsers.py:2410]: logits shape is [batch, num_classes]. It is the output of inference.

Comment (runwangdl) [reply Deeploy/Targets/Generic/Parsers.py:2410]: Logits shape is [batch, num_classes]. It is the output of inference.

File: Deeploy/Targets/Generic/Parsers.py (hunk 3 above)
Comment (runwangdl) [reply Deeploy/Targets/Generic/Parsers.py:2442]: log_prob has the shape same as logits which is [batch, num_classes]

File: Deeploy/Targets/Generic/Parsers.py (hunk 2 above)
Comment (runwangdl) [reply Deeploy/Targets/Generic/Parsers.py]: Fixed in [a5a330b](https://github.com/pulp-platform/Deeploy/pull/53/commits/a5a330b501d0f52e73b49a8670abb461ac092cd6)

File: Deeploy/Targets/PULPOpen/Templates/SoftmaxCrossEntropyLossTemplate.py (hunk 4 above)
Comment (runwangdl) [reply Deeploy/Targets/PULPOpen/Templates/SoftmaxCrossEntropyLossTemplate.py]: Fixed in [a5a330b](https://github.com/pulp-platform/Deeploy/pull/53/commits/a5a330b501d0f52e73b49a8670abb461ac092cd6)

File: .github/workflows/CI.yml (hunk 5 above)
Comment (runwangdl) [reply .github/workflows/CI.yml]: Fixd in [a5a330b](https://github.com/pulp-platform/Deeploy/pull/53/commits/a5a330b501d0f52e73b49a8670abb461ac092cd6)

**Comment by runwangdl** [comment]
Conflicts merged in [10e4cc0](https://github.com/pulp-platform/Deeploy/pull/53/commits/10e4cc0f873309ff1bc6a3e189cce0a863b3e6a5)

**Comment by runwangdl** [comment]
Conflicts solved.
